qiskit/optimization/algorithms/warm_start_qaoa_optimizer.py

```diff
@@ -0,0 +1,258 @@
+                :class:`~qiskit.optimization.converters.QuadraticProgramToQubo` will be used.
+
+        Raises:
+            AquaError: if ``epsilon`` is not specified for the warm start QAOA.
+        """
+        if epsilon is None:
+            raise AquaError('Epsilon must be specified for the warm start QAOA')
+
+        self._pre_solver = pre_solver
+        self._num_relaxed_solutions = num_relaxed_solutions
+        self._qaoa = qaoa
+        self._epsilon = epsilon
+        self._aggregator = aggregator
+        super().__init__(qaoa, penalty, converters)
+
+    def get_compatibility_msg(self, problem: QuadraticProgram) -> Optional[str]:
+        """Checks whether a given problem can be solved with this optimizer.
+
+        Checks whether the given problem is compatible, i.e., whether the problem can be converted
+        to a QUBO, and otherwise, returns a message explaining the incompatibility.
+
+        Args:
+            problem: The optimization problem to check compatibility.
+
+        Returns:
+            A message describing the incompatibility.
+        """
+        return super().get_compatibility_msg(problem)
+
+    def solve(self, problem: QuadraticProgram) -> OptimizationResult:
+        """Tries to solves the given problem using the optimizer.
+
+        The pre-solver is run to warm-start the solver. Next, the optimizer is run
+        to try to solve the optimization problem.
+
+        Args:
+            problem: The problem to be solved.
+
+        Returns:
+            The result of the optimizer applied to the problem.
+
+        Raises:
+            QiskitOptimizationError: If problem not compatible.
+        """
+
+        msg = self.get_compatibility_msg(problem)
+        if len(msg) > 0:
+            raise QiskitOptimizationError('Incompatible problem: {}'.format(msg))
+
+        # convert problem to QUBO
+        # qubo_converter = QuadraticProgramToQubo(self._penalty)
+        # qubo_problem = qubo_converter.convert(problem)
+        qubo_problem = self._convert(problem, self._converters)
+
+        # construct operator and offset
+        operator, offset = qubo_problem.to_ising()
+
+        opt_result = self._pre_solver.solve(problem)
+        num_relaxed_solutions = min(self._num_relaxed_solutions, len(opt_result.all_solutions))
+        # todo: accessing all solutions
+        relaxed_solutions = opt_result.all_solutions[:num_relaxed_solutions]
+
+        results = []  # type: List[MinimumEigenOptimizationResult]
+        for relaxed_solution, value in relaxed_solutions:
+            # Set the solver using the result of the pre-solver.
+            initial_variables = self._create_initial_variables(relaxed_solution)
+            # todo: no such setter
+            self._qaoa.initial_state = self._create_initial_state(initial_variables)
+            # todo: no such setter
+            self._qaoa.mixer = self._create_mixer(initial_variables)
```

Comment:
How can I provide my own mixer to Warm-start QAOA? E.g. in the WS paper we modified the mixer to deal with GW.

Reply:
Added `MixerFactory`.

Reply:
Now, there's a circuit factory and both initial state and mixer can be created by a user.